crates/handlers/src/views/recovery/finish.rs

```diff
@@ -0,0 +1,246 @@
+    }
+
+    let user_email = repo
+        .user_email()
+        .lookup(ticket.user_email_id)
+        .await?
+        // Only allow confirmed email addresses
+        .filter(|email| email.confirmed_at.is_some())
+        .context("Unknown email address")?;
+
+    let user = repo
+        .user()
+        .lookup(user_email.user_id)
+        .await?
+        .context("Invalid user")?;
+
+    if !user.is_valid() {
+        // TODO: render a 'account locked' page
+        let rendered = templates.render_error(
+            &ErrorContext::new()
+                .with_code("Account locked")
+                .with_language(&locale),
+        )?;
+        return Ok((cookie_jar, Html(rendered)).into_response());
+    }
+
+    let context = RecoveryFinishContext::new(user)
+        .with_csrf(csrf_token.form_value())
+        .with_language(locale);
+
+    let rendered = templates.render_recovery_finish(&context)?;
+
+    Ok((cookie_jar, Html(rendered)).into_response())
+}
+
+pub(crate) async fn post(
+    mut rng: BoxRng,
+    clock: BoxClock,
+    mut repo: BoxRepository,
+    mut policy: Policy,
+    State(password_manager): State<PasswordManager>,
+    State(templates): State<Templates>,
+    State(url_builder): State<UrlBuilder>,
+    PreferredLanguage(locale): PreferredLanguage,
+    cookie_jar: CookieJar,
+    Query(query): Query<RouteQuery>,
+    Form(form): Form<ProtectedForm<RouteForm>>,
+) -> Result<Response, FancyError> {
+    let (csrf_token, cookie_jar) = cookie_jar.csrf_token(&clock, &mut rng);
+
+    let ticket = repo
+        .user_recovery()
+        .find_ticket(&query.ticket)
+        .await?
+        .context("Unknown ticket")?;
+
+    let session = repo
+        .user_recovery()
+        .lookup_session(ticket.user_recovery_session_id)
+        .await?
+        .context("Unknown session")?;
+
+    if !ticket.active(clock.now()) || session.consumed_at.is_some() {
+        // TODO: render a 'link expired' page
+        let rendered = templates.render_error(
+            &ErrorContext::new()
+                .with_code("Link expired")
+                .with_language(&locale),
+        )?;
+        return Ok((cookie_jar, Html(rendered)).into_response());
+    }
+
+    let user_email = repo
+        .user_email()
+        .lookup(ticket.user_email_id)
+        .await?
+        // Only allow confirmed email addresses
+        .filter(|email| email.confirmed_at.is_some())
+        .context("Unknown email address")?;
+
+    let user = repo
+        .user()
+        .lookup(user_email.user_id)
+        .await?
+        .context("Invalid user")?;
+
+    if !user.is_valid() {
+        // TODO: render a 'account locked' page
+        let rendered = templates.render_error(
+            &ErrorContext::new()
+                .with_code("Account locked")
+                .with_language(&locale),
+        )?;
+        return Ok((cookie_jar, Html(rendered)).into_response());
+    }
```

Comment:
all the logic above this seems to be in common with the `get` route; wonder if can we pull this out into a separate function without it being too awkward?

Reply:
any insight why this is hard? what happens if you try it?

Reply:
ok for now